### PR TITLE
DOC: Fix documentation for np.dtype.kind to include 'T' for StringDType

### LINKED ISSUE #27948: DOC: dtype kind 'T' is not documented
### Issue with current documentation:

This page https://numpy.org/doc/stable/reference/generated/numpy.dtype.kind.html#numpy.dtype.kind explicitly lists all possible returns from `np.dtype.kind` as 
> one of 'biufcmMOSUV'

However, the character associated with `StringDType` (`T`) isn't listed
```python
>>> import numpy as np
>>> np.dtypes.StringDType().kind
'T'
```

### Idea or request for content:

_No response_

### PR DESCRIPTION
**Pull request by Uvi-12**
Fixes #27948 

This PR updates the NumPy documentation for the `np.dtype.kind` attribute to include 'T' for `StringDType`.

### CODE REVIEW

**Review by StanFromIreland — COMMENTED** [review]
Seems all good in my opinion!

**Comment by Uvi-12** [comment]
@StanFromIreland Thank you for the review. If everything looks good, please proceed with merging the PR.

File: numpy/_core/_add_newdocs.py [hunk 1]
```diff
@@ -6095,6 +6096,9 @@
     >>> dt = np.dtype([('field1', 'f8')])
     >>> dt.kind
     'V'
+    >>> dt = np.dtype('U10')  # Example for StringDType
+    >>> dt.kind
+    'T'
```
Comment (seberg) [inline numpy/_core/_add_newdocs.py:6101]: This example doens't compute, not sure why it's not tested...  Maybe just delete it even?

**Comment by Uvi-12** [comment]
@seberg Thank you for the review, I have made changes according to it. If everything looks good, please proceed with merging the PR.

File: numpy/_core/_add_newdocs.py (hunk 1 above)
Comment (mattip) [reply numpy/_core/_add_newdocs.py]: @ev-br any idea why doctests are not running here?

Comment (mattip) [reply numpy/_core/_add_newdocs.py]: Running `spin check-docs -v` and checking for dtype(s) gives me
```
$ spin check-docs -v | grep "\<dtypes*\>"
numpy/__init__.py::numpy.dtype PASSED                                    [ 13%]
numpy/__init__.py::numpy.dtype.__class_getitem__ PASSED                  [ 13%]
numpy/__init__.py::numpy.dtype.newbyteorder PASSED                       [ 13%]
numpy/dtypes.py::numpy.dtypes.StringDType PASSED                         [ 18%]
numpy/matlib.py::numpy.matlib.dtype PASSED                               [ 31%]
numpy/matlib.py::numpy.matlib.dtype.__class_getitem__ PASSED             [ 31%]
numpy/matlib.py::numpy.matlib.dtype.newbyteorder PASSED                  [ 31%]
numpy/core/multiarray.py::numpy.core.multiarray.dtype PASSED             [ 43%]
numpy/core/multiarray.py::numpy.core.multiarray.dtype.__class_getitem__ PASSED [ 43%]
numpy/core/multiarray.py::numpy.core.multiarray.dtype.newbyteorder PASSED [ 43%]
numpy/core/numeric.py::numpy.core.numeric.dtype PASSED                   [ 46%]
numpy/core/numeric.py::numpy.core.numeric.dtype.__class_getitem__ PASSED [ 46%]
numpy/core/numeric.py::numpy.core.numeric.dtype.newbyteorder PASSED      [ 46%]
```

I wonder how we get collection to see `numpy._core.multiarray.dtype`

Comment (Uvi-12) [reply numpy/_core/_add_newdocs.py]: Thanks, @mattip, I noticed that the doctest example for np.dtype.kind wasn’t running during the tests. Based on @seberg’s suggestion, I removed the example- dt = np.dtype('U10') as it didn't compute. Please let me know if there are any further changes to be made.

Comment (ev-br) [reply numpy/_core/_add_newdocs.py]: Hmmm, interesting. So dtype _methods_ are being collected but dtype `attributes` are not?

```
$ spin check-docs -v -- -k'dtype'
... snip ...
$ export PYTHONPATH="/home/br/repos/numpy/build-install/usr/lib/python3.11/site-packages"
$ /home/br/miniforge3/envs/numpy-dev/bin/python3.11 -P -c 'import numpy'
$ cd /home/br/repos/numpy/build-install/usr/lib/python3.11/site-packages
$ /home/br/miniforge3/envs/numpy-dev/bin/python3.11 -P -m pytest -v -m 'not slow' -kdtype --doctest-modules --doctest-collect=api
=========================================================================================== test session starts ===========================================================================================
platform linux -- Python 3.11.9, pytest-8.2.2, pluggy-1.5.0 -- /home/br/miniforge3/envs/numpy-dev/bin/python3.11
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase(PosixPath('/home/br/repos/numpy/build-install/usr/lib/python3.11/site-packages/.hypothesis/examples'))
rootdir: /home/br/repos/numpy
configfile: pytest.ini
plugins: cov-5.0.0, scipy_doctest-1.5.1, xdist-3.6.1, hypothesis-6.103.2
collected 3947 items / 2664 deselected / 1283 selected                                                                                                                                                    

numpy/__init__.py::numpy.isdtype PASSED                                                                                                                                                             [  4%]
numpy/__init__.py::numpy.dtype PASSED                                                                                                                                                               [  8%]
numpy/__init__.py::numpy.dtype.__class_getitem__ PASSED                                                                                                                                             [ 12%]
numpy/__init__.py::numpy.dtype.newbyteorder PASSED                                                                                                                                                  [ 16%]
numpy/__init__.py::numpy.ufunc.resolve_dtypes PASSED                                                                                                                                                [ 20%]
numpy/__init__.py::numpy.issubdtype PASSED                                                                                                                                                          [ 24%]
numpy/core/multiarray.py::numpy.core.multiarray.dtype PASSED                                                                                                                                        [ 28%]
numpy/core/multiarray.py::numpy.core.multiarray.dtype.__class_getitem__ PASSED                                                                                                                      [ 32%]
numpy/core/multiarray.py::numpy.core.multiarray.dtype.newbyteorder PASSED                                                                                                                           [ 36%]
numpy/core/numeric.py::numpy.core.numeric.ufunc.resolve_dtypes PASSED                                                                                                                               [ 40%]
numpy/core/numeric.py::numpy.core.numeric.dtype PASSED                                                                                                                                              [ 44%]
numpy/core/numeric.py::numpy.core.numeric.dtype.__class_getitem__ PASSED                                                                                                                            [ 48%]
numpy/core/numeric.py::numpy.core.numeric.dtype.newbyteorder PASSED                                                                                                                                 [ 52%]
numpy/core/numeric.py::numpy.core.numeric.issubdtype PASSED                                                                                                                                         [ 56%]
numpy/core/numeric.py::numpy.core.numeric.isdtype PASSED                                                                                                                                            [ 60%]
numpy/core/numerictypes.py::numpy.core.numerictypes.issubdtype PASSED                                                                                                                               [ 64%]
numpy/core/numerictypes.py::numpy.core.numerictypes.isdtype PASSED                                                                                                                                  [ 68%]
numpy/dtypes.py::numpy.dtypes.StringDType PASSED                                                                                                                                                    [ 72%]
numpy/exceptions.py::numpy.exceptions.DTypePromotionError PASSED                                                                                                                                    [ 76%]
numpy/matlib.py::numpy.matlib.isdtype PASSED                                                                                                                                                        [ 80%]
numpy/matlib.py::numpy.matlib.dtype PASSED                                                                                                                                                          [ 84%]
numpy/matlib.py::numpy.matlib.dtype.__class_getitem__ PASSED                                                                                                                                        [ 88%]
numpy/matlib.py::numpy.matlib.dtype.newbyteorder PASSED                                                                                                                                             [ 92%]
numpy/matlib.py::numpy.matlib.ufunc.resolve_dtypes PASSED                                                                                                                                           [ 96%]
numpy/matlib.py::numpy.matlib.issubdtype PASSED                                                                                                                                                     [100%]
```

Is there a difference between how `dtype.newbyteorder` is handled/attached/exposed to python and `dtype.kind`?

Comment (Uvi-12) [reply numpy/_core/_add_newdocs.py]: Hi @ev-br, @mattip, I’m new to contributing to NumPy, and I’m wondering if the issue with doctests not running is related to the documentation changes I made, or is it due to something else? Please let me know.

Comment (mattip) [reply numpy/_core/_add_newdocs.py]: Not due to anything you did, but by touching this part of the docstring you exposed a shortcoming in our testing. These unintended consequences happen often in NumPy, so we take advantage of their unexpected appearances and try to work out what is going on.

Comment (ev-br) [reply numpy/_core/_add_newdocs.py]: Actually, I sort of see what's going on here.

TL;DR: looks like the standard library `doctest` module does not recognize what numpy does when attaching `np.dtype` members.

Dropping a breakpoint into https://github.com/scipy/scipy_doctest/blob/main/scipy_doctest/frontend.py#L99 and running `$ spin check-docs -- -k'dtype' --collect-only`, and stepping through all the way to doctest finding, we find deep in the standard library code: 

```
> /home/br/miniforge3/envs/numpy-dev/lib/python3.11/doctest.py(1040)_find()
-> if isinstance(val, (staticmethod, classmethod)):
(Pdb) p val
<member 'kind' of 'numpy.dtype' objects>
(Pdb) p type(val)
<class 'member_descriptor'>
(Pdb) l
1035 	
1036 	        # Look for tests in a class's contained objects.
1037 	        if inspect.isclass(obj) and self._recurse:
1038 	            for valname, val in obj.__dict__.items():
1039 	                # Special handling for staticmethod/classmethod.
1040B->	                if isinstance(val, (staticmethod, classmethod)):
1041 	                    val = val.__func__
1042 	
1043 	                # Recurse to methods, properties, and nested classes.
1044 	                if ((inspect.isroutine(val) or inspect.isclass(val) or
1045 	                      isinstance(val, property)) and
(Pdb) p inspect.isroutine(val)
False
(Pdb) p inspect.isclass(val)
False
(Pdb) p isinstance(val, property)
False
```

So, `np.dtype.kind` is an instance of `class <member_descriptor>`. Whatever this is, it's not recognized as a routine or a property, and thus its docstring is skipped. 

Conculsion: the problem is in the standard library `doctest`, it's been there forever (tm), is likely present in any tool which relies on or extends `doctest`, can be worked around in `scipy-doctest` for numpy specifically (and whoever else is doing similar sorts of black magic attaching methods/attributes to C-defined objects).  
Opened https://github.com/scipy/scipy_doctest/issues/178 to track this.

Comment (mattip) [reply numpy/_core/_add_newdocs.py]: I opened #28002

**Review by mattip — APPROVED** [review]
LGTM. The doctest problem is not connected to this PR.

**Comment by Uvi-12** [comment]
Thank you @StanFromIreland @seberg @mattip @ev-br for the reviews and helping me out through the PR.
